### ISSUE TITLE
Table redesign

### ISSUE THREAD

**Issue by kevee** [COVID19Tracking/website#457]
Meta issue of #426, #317, and #379, which have been closed and merged into this one:

We need a flexible way to handle tables across all devices. Let's create a new table component that is:

- **Responsive** - Stacks and reorients itself in mobile devices (#317 has lots of ideas)
- **Customizable** -  users can add/remove columns or column groups
- **Stable for linking** - Customizations by the user should be preserved for easy sharing/linking

This table component will first be applied to:

- [US Historical Data](https://covidtracking.com/data/us-daily)
- [State history views](https://covidtracking.com/data/state/california)

**Comment by toddparker** [comment]
Ok gang let’s get a conversation on what we think is a good approach. In general, the most popular patterns for responsive tables are:

**Overflow** - just wrap it with a side scrolling container (current site does this). This is obviously the easiest approach but since the first cell scrolls away, you lose the context of what each row is so it can be disorientating and you need to scroll around a lot. Truly fixing the first column to keep context may still not be possible with CSS(?) but we may be able to do some hackery to keep it fixed (accessibility and layout robustness are key considerations) - https://github.com/COVID19Tracking/website/issues/317#issuecomment-607613708

**Stack** - re-format the table so each cell becomes a stacked ‘row’ and the column headers become labels. This is best for situations where comparing data between rows isn’t important because the stacked style means that values are physically far apart. The main advantage is everything is visible on-screen:  no need to scroll or toggle columns (content complexity is a consideration, lowt of rows and columns will stack out to a very long page)  - https://github.com/COVID19Tracking/website/issues/317#issuecomment-609918295

**Responsive columns** - prioritize the columns based on their relative importance and hide less important columns as the viewport shrinks so the table fits on-screen and ideally provide a way to let people override these defaults and pick which columns are most important to them. This mode keeps the context (usually the first cell with the name) and allows for easy row-by-row comparisons at the expense of seeing all the columns (discovery of the column picker is a key consideration) - https://github.com/COVID19Tracking/website/issues/317#issuecomment-610467846

**Collapsibles** - some of these ideas could be combined with expanding/collapsing content on demand. For example, we could show a simple table with just a few columns on a phone (or just the state name) and expand to show a stacked presentation of all the data values on demand. This can help manage the complexity of many columns of data presented in stack mode so you can understand the lay of the land - https://github.com/COVID19Tracking/website/issues/317#issuecomment-609414661

**Comment by kevee** [comment]
Wow, @toddparker, this is amazing! My thoughts:

I'm super biased, but I love the idea of not using tables when there is a single row of information - it's a lot for a screen reader user to tab around, and it's difficult to scan for sighted users. I'm going to get some more eyes on this.

**Comment by beep** [comment]
oh my goodness @toddparker, this is _stellar_. I think all the adaptation cases you laid out make sense. (But what I _really_ like is that I think you’re describing a really solid vocabulary we could pull from, which’d let us make decisions about when and how different kinds of tables should adapt.)

I’ve used tablesaw more than a couple times, and _love_ it. I gather there were some performance issues around using it with gatsby, though—do we know if that’ll be a blocker here?

I’m not aware of any a11y/usability issues with tables that have a single body row, but _also_:

- If we can simplify the information we’re displaying, I’d totally be up for rethinking the table layout, and/or using a different markup structure.
- This makes me think it’d be _cool_ if the site had an accessibility statement? It’d be great to be able to get accessibility feedback directly from our users.

**Comment by kevee** [comment]
The tablesaw thing was done in haste, I'll take that out and see if I can't address the performance issues.

If we really like the stuff that tablesaw does, and we can't get it to work, we can still reproduce it within React.

100% agree on accessibility statement, I'm making an issue for that. In the past in testing with screen reader users (and me personally, but I am *not* a regular SR user) dropping into table mode for one row is just a bit annoying.

**Comment by toddparker** [comment]
Thanks for the positive feedback so fa! Once we make some decisions of which patterns we want to pursue, I could pair up with a developer to help guide the tablesaw work and get some of the technical risk out of the way with a POC to make sure performance and Gatsby/React compat aren't going to be an issue. I see us using this plugin at most a few times per page, wondering is the original issue was trying to spin it up for 50+ tables in stack mode because there is a lot of DOM manipulation happening that might have made React mad. If we can consolidate the state data into one table with the column picker mode, that should sidestep that particular issue. I can work @kevee on that, or anyone who has time.@jefflembeck mentioned he might have time to help and is familiar with tablesaw.

I'd love to get some design help on visualizing how we'd like to present these different kinds of tables. 
- Seems we could whip of some mockups for the state summary info that builds on @kevee's non-table idea. 
- We also could think about how we might want to style up our multi-row table with the column pickers to make it clearer the you can choose columns. 

Maybe I can convince @beep and some other of our stellar designer to help on that front...

**Comment by toddparker** [comment]
**Just parking a rough draft of this here that I'll be adding to. We should wait to hear back on the larger direction before we do more work...**

It would be nice if we had a coherent family of visual styles that unified these tables even if they have different modes. If folks want to help out with something concrete, I've started sketching out some design tasks to move the conversation forward:

## Redesign state summary tables
It's clear that we have these single row tables that provide a snapshot of the current situation per state. Let's sketch out how we'd like to re-imagine the top of these pages, either with a stack table or a custom label/value pair setup that quickly communicates the data. 

- [ ] This doesn't need to be structured as a table. It can be formatted text
- [ ] Must be responsive and show all the data neatly on small screens
- [ ] Typography and styles should fit into the larger design direction
- [ ] Must be accessible (more of a tech consideration but things should be clearly labeled/associated)

Basically, seems everything below the state name and before the history table is fair game:
https://covidtracking.com/data/state/new-york

![image](https://user-images.githubusercontent.com/359929/79007888-323e2d80-7b2a-11ea-90ba-03189f810b51.png)

There are also some links and notes from the New York section on the [State-by-state](https://covidtracking.com/data) page that are also on the detail page but presented slightly differently. Here is the whole NY block for comparison:

![image](https://user-images.githubusercontent.com/359929/79008059-906b1080-7b2a-11ea-9db8-ea4c3ac64d5c.png)

Earlier sketch showing this more as formatted text than a table

![](https://user-images.githubusercontent.com/359929/78927069-373ca780-7a6c-11ea-8acb-6382c119c9c4.png)



## Redesign multi-row tables to include responsive columns and column picker

Some things to think about:
- [ ] Fonts and sizing for TH and TDs
- [ ] Do we use row strokes, alternating stripes?
- [ ] Should headers have a bg or strokes? For 2 level headings what is the typography?
- [ ] Do we have strokes for each cell? Maybe just for the right edge or grouped headers?
- [ ] How should the column selection work? Can we make the affordance for that super clear?
- [ ] Think about sorting feedback/affordance
- [ ] Clean up our heading structure for tables with two level headings (useful if we combine all the states into one table for the state-by-state page)
- [ ] For each table, we need to start thinking about prioritizing columns so they can drop off responsively 

### [US Historical Data](https://covidtracking.com/data/us-daily)
![image](https://user-images.githubusercontent.com/359929/78927396-d497db80-7a6c-11ea-9a8a-43d11da0da50.png)

### [State detail - historical table](https://covidtracking.com/data/state/alabama#historical)
![image](https://user-images.githubusercontent.com/359929/78927486-fd1fd580-7a6c-11ea-89a8-d56faf7428c0.png)

**Comment by kevee** [comment]
I got tablesaw to work in our site and just [tossed it on the US history page in this staging site](https://table-saw-test--goofy-tesla-43452a.netlify.com/data/us-daily). Looks like it works just fine.

There *is* a performance hit, Lighthouse gives the page with tablesaw a 64 for performance, while our current live site is 96. There's a noticeable pause between page load and the tablesaw loading, but I am sure we can get that time down a bit.

**Comment by toddparker** [comment]
Hey that's great @kevee. We can take at look at the performance to see if we have any recs. One thing you're going to want to add is priorities for each column so they can drop off progressively.
https://github.com/filamentgroup/tablesaw

You can see an example below but the idea is you add a `data-tablesaw-priority="2"` to the column header cells in the table. Use a value of `persist` for the first column (date) so it never hides then add priority values of 1-6 where `1` will come in on smallest screens and `6` will come in on widest screens. It's ok to have multiples of a number.

https://jsbin.com/zaxuhoz/edit

![image](https://user-images.githubusercontent.com/359929/79016880-09279800-7b3e-11ea-93f4-42c64e30b5f8.png)

**Comment by scottjehl** [comment]
Great discussion here. If it helps,  here are some links to performance runs to see what we're looking at. 

## Pagespeed Insights

Here's a Pagespeed Insights run of that page. It seemed to do well in both mobile and desktop. I agree there's a little delay in time to interactive, but at least in this run it looked pretty solid. 92 On mobile and 100 on desktop. Nice work.

https://developers.google.com/speed/pagespeed/insights/?url=https%3A%2F%2Ftable-saw-test--goofy-tesla-43452a.netlify.com%2Fdata%2Fus-daily&tab=mobile
<img width="1016" alt="image" src="https://user-images.githubusercontent.com/214783/79025474-7d1c6d00-7b4b-11ea-9a9f-cb024e8f5382.png">

In the summary it looks like the main complaint that PSI has is in time spent waiting for blocking scripts to load (a third party netlify script). That may be an artifact of loading from this test domain, however. 

## Webpage Test

WPT gives a better look at what's going on.  There are a few seconds where the CPU is working hard on nothing but JS, but it's at least after render. I agree that there could be ways to speed that up too. 

## Redirect Perf Issue...

One unrelated thing to note on that demo page is there's a 301 redirect that takes the first second and a half of the 3 seconds to start render, and it looks like it's only there to add a trailing slash and then remove it... I think? 

I checked the live site and that redirect is there too, even on refresh. I'm not sure if it's intentional, but if that 301 could be dropped, the page could render almost 2 seconds earlier on 3G.
<img width="661" alt="image" src="https://user-images.githubusercontent.com/214783/79026446-7d6a3780-7b4e-11ea-9fab-8fd938de64df.png">

https://www.webpagetest.org/result/200410_W2_b611f61d9b39e6b1023b60001f0f3c65/2/details/#waterfall_view_step1

Sorry for the small tangent away from the tablesaw stuff. If this last point is useful I'm happy to file a separate issue.

**Comment by kevee** [comment]
Hi, @scottjehl! Yes, if you can file a separate issue with info on the trailing slash, I think that's a simple issue on Netlify. Thanks for all this insights!

**Comment by toddparker** [comment]
I think we can work on this in a discrete steps instead of a big bang kind of situation. I'l;l make separate issues so we can keep things focused.

* improve general table styles and accessibility 
* add responsive modes, maybe in a few releases: stack and column chooser (flat vs. grouped columns)
* enhancements to add things like spark lines and support new data columns/types
* custom state summary table (once that content strategy work is done)

**Comment by kevee** [comment]
To track: #792
